### DISCUSSION TITLE
Random Generator and AT elements

### DISCUSSION THREAD

**Discussion by oscarxblanco** [atcollab/at#879] [Q&A] [answered]
Dear all, 
I have several questions about random generators in AT, you might have already discussed.
I see that AT uses a C implementation of the random generator https://www.pcg-random.org/using-pcg-c.html saved in the `parameters` struct. Two streams are created `common_rng` and `thread_rng`.
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/atintegrators/attypes.h#L24-L25
I will separate the questions in three parts : PYAT, AT MATLAB, and Elements using random numbers.
Please, let me know what I got wrong about how it works.

---
**PYAT**
Both streams are initialized in pyat using the C function `pcg32_srandom_r`, where `AT_RNG_ING`, `seed` and `rank` are set before tracking using the static constants and the MPI settings. So, there is no way the user could set the two values used for initialization of each stream, right ?
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/pyat/at.c#L799-L800
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/atintegrators/atrandom.c#L28
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/pyat/at/tracking/track.py#L43
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/pyat/at.c#L787-L793

---
**AT MATLAB**
In AT MATLAB it seems that they are only initialized once in the MATLAB session (not on every track as in pyat) and it uses static values, right ?
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/atmat/attrack/atpass.c#L73-L74
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/atintegrators/atrandom.c#L30-L31

---
**Elements using random numbers**
There are a few elements that use random numbers: `atQuantDiff` and `atVariableMultipole`.
Both state that a seed could be used. However, the QuantDiffPass method uses the `Param->thread_rng` stream and the VariableThinMultipole  uses a static  initializer `PCG32_INITIALIZER` inside  `atrandn`, but, we just saw that the user has no control over the seed, so, the user seed is actually never used, right ?

(QuantDiff)
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/atmat/lattice/element_creation/atQuantDiff.m#L17-L19
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/atintegrators/QuantDiffPass.c#L59
(VariableThinMultipole)
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/pyat/at/lattice/variable_elements.py#L49-L50
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/atmat/lattice/element_creation/atvariablemultipole.m#L22
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/atintegrators/VariableThinMPolePass.c#L69
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/atintegrators/atrandom.c#L153-L156
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/atintegrators/atrandom.c#L46
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/atintegrators/atrandom.c#L42-L44
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/atintegrators/atrandom.c#L27-L28

---
Moreover, and assuming all I asked before is correct. The C implementation of the generator seems to be compatible with pyat through numpy https://numpy.org/doc/2.0/reference/random/bit_generators/pcg64.html , but, I couldn't find the same for Matlab. I guess it means that the generator can not belong to the element because it needs to work in both python and matlab, therefore, it needs to be kept at the C level.

If one desires to use the element seed in tracking, a new stream should be initialized before starting tracking, which means checking the ring for random elements and create every stream.

Alternatively, one could forget about the element seed and create an input for the user seed (`initstate`,`initseq`) to initialize the streams in the `parameter` struct before the tracking, and make sure the elements use only one of the two streams (`common_rng` or `thread_rng`). The element might be modified to use one or the other.

A third solution would be to remove any `seed` input on the element. It would only change the documentation and help messages.

One other option would be getting PCG to work inside MATLAB, so, the element could have its own stream when defined, and the tracking would only the element own stream.

At last, maybe a more complicated way, one could change to a generator that exist in both python and matlab, and again implement the stream when creating the element.

**Comment by lfarv** [comment, marked as answer]
I'll put things in another way:  random generators used in python and C are distinct, and as far as I can see, there is no specific random generator in Matlab, only the Matlab standard ones. The motivation is to avoid having the C engine depending on any external library. The whole implementation is in atrandom.c, and the behaviour is identical when running under python or Matlab. the C generators can be accessed from python but not from Matlab. For element passmethods, the generators which matter are the C generators, and the will work for both Matlab and python. There are 2 streams:

- the "thread" stream `Param->thread_rng` will generate different streams in different parallel threads, either for python parallel execution, MPI or OpenMP. This is the one used in QuantumDiff, since particles tracked in different streams must be independent.
- the "common" stream `Param->common_rng` will generate the same stream for different parallel threads. This one should be used in variable elements for instance, so that particles tracked in different parallel threads experience the same element.

The buffers for these streams are defined in at.c:
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/pyat/at.c#L73-L75

There is a third stream using a static private buffer, with simpler function calls without the buffer argument. Its buffer is defined here:
https://github.com/atcollab/at/blob/8620504680c7294ae645280379adea62160bca51/atintegrators/atrandom.c#L46
It should be used only in regions without parallel processing.

There is no provision to have a specific stream for each element: element random values should be taken from one of the 2 tracking streams. So you are right, the seed specified in the variable multipole is not used. But the present implementation in `VariableThinMPolePass` is wrong: on should use `atrandn_r(Param->common_rng, 0.0, 1.0)` instead of `atrandn(0.0, 1.0)`. As it is, particles tracked in parallel threads will not see the same multipole.

Now comes the question of explicitly seeding the generators. It is possible in python, but not in Matlab. For the moment, a python random generator is used to provide the seed of C generators at the beginning of each tracking. It would be easy to add an optional `seed` argument to `Lattice.track` to allow a user-defined seed. If I understand well, that's what you propose here:

> Alternatively, one could forget about the element seed and create an input for the user seed (initstate,initseq) to initialize the streams in the parameter struct before the tracking, and make sure the elements use only one of the two streams (common_rng or thread_rng).

That's more difficult in Matlab: the tracking streams must be declared in atpass.c to be accessed by element passmethods. Then they cannot be seen by other mex-functions allowing explicit seeding. Defining them in a shared library accessed by several Mex-function is a solution, but how to compile and link a shared library with the Mex command?

Finally about the "python" generators defined in lattice.options.py: they are completely independent form the C ones, and I don't think they are used anywhere except seeding the C generators.

**Reply by oscarxblanco** [reply]
Dear @lfarv , thank you for the answer it is clearer now.

From what I see, there is no need to add another stream right now.
I will remove the seed parameter from the VariableThinMultipole element that I have been editing.
Should I also remove the seed parameter info in the help messages of QuantDiff ?

And one more question about `atrandn_r(Param->common_rng, 0.0, 1.0)`. Would it be acceptable to pass a mean and std defined by the user ? i.e. `atrandn_r(Param->common_rng, user_mean, user_std)` ?

**Reply by lfarv** [reply]
> Would it be acceptable to pass a mean and std defined by the user ? i.e. atrandn_r(Param->common_rng, user_mean, user_std) ?

Of course you can, the function is built for that. It depends on what you want to do.

**Comment by carmignani** [comment]
Hello,
I did not follow this discussion, but today I did some tracking in the cluster with compiled matlab and with the quantum excitation on and I noticed that the random number generator seed was the same in each of the processes. So I wanted to change the seeds, but I noticed that the seed parameter in the quantum diffusion element has been removed.
Is there a way now to give a different seed to each process?
The field "seed" as we had a few years ago in the QuantDiffPass was not good?

**Comment by oscarxblanco** [comment]
Hi Nicola,
I have gone through the modifications on `QuantDiff.c`, and the removal of the seed dates back to Dec,5 2022.
![image](https://github.com/user-attachments/assets/17ceabbf-ae5b-4433-a031-3435df748052)
https://github.com/atcollab/at/commit/8ef93db08b7a7c0fc0848aac91385e928c70241f#diff-cb146f57a02faf5a944b035fc9e297c5f528e322d1ecd9a801b856f8a13602d2

This commit mentions an update of the pass function that does not longer takes a seed.
![image](https://github.com/user-attachments/assets/8c6fd864-2b5c-4928-9f66-5d5ef652cd30)
And there are several commits before and after referring to fixing the compilation with OpenMP.

The current version does not take `Seed` as argument and when called from matlab it uses a global generator described by @lfarv above.
`static pcg32_random_t pcg32_global = PCG32_INITIALIZER`

**Comment by carmignani** [comment]
Thanks @oscarxblanco. I think we need a way to pass a seed to the c code from matlab, for when we run the same identical process many times. In the past it was possible to specify a different seed in the quantum diffusion element for each process, but now it is not possible.
Can't we have a seed input in ringpass for example?

**Reply by oscarxblanco** [reply]
Maybe @lfarv or @swhite2401 can answer that ?
I don't know enough about those modification to give an answer right away.

**Comment by swhite2401** [comment]
The seed can be set in python through a dedicated function, we had a quick look this morning with Nicola but could no find the matlab equivalent. I don't whether it is easy to implement but it should be possible maybe it even exist... we have to wait for @lfarv.

A short description of random generators in the documentation would be helpful as well.

**Comment by lfarv** [comment]
Hi all. Here are some explanations, in addition to the previous ones:

The random generator has been taken out of elements because of parallel processing. In the case of python multiprocessing and MPI, the generators would be copied and would generate the same sequence. That's why there are now 2 different streams: a common one and a thread-specific one, handled  in the C "main" function, `atpass.c` for Matlab and `at.c` for python.

Resetting the seed can be done in python because we can have several entry points into the same `atpass.c` C module: one entry `atpass` for tracking, and another one `reset_rng` for seeding the generators. That is not possible in Matlab because mex-functions have a single entry `MexFunction` and are independent of each other.

So the only solution in Matlab would be to add another input argument to the `atpass` call, to optionally give a seed when starting the tracking. This was also mentioned [above](https://github.com/atcollab/at/discussions/879#discussioncomment-11615978) for the python `Lattice.track` method, and could also be done for the Matlab `atpass` function.

The problem is tricky because of the multiple parallel processing options: OpenMP for both python and Matlab, python multiprocessing and MPI for python only, and Matlab parallel-processing toolbox. The additional argument on tracking startup should work for all except Matlab parallel toolbox, though the generated sequence will be different for different methods.
If yo agree, I'll open a pull request for adding a new argument to `Lattice.track` and `atpass`.

**Reply by swhite2401** [reply]
Sounds good to me! Would it be possible to write a bit of documentation as well? Thanks!

**Reply by carmignani** [reply]
An additional input to atpass in matlab would be perfect for me.
Can we use OpenMP from matlab?

**Reply by lfarv** [reply]
OK, let's go for that.

> Can we use OpenMP from matlab?

Yes:
`>> atmexall -openmp`

**Reply by carmignani** [reply]
Is there some documentation for openMP in matlab AT? I was not aware about it. 
Is the parallel toolbox needed?

**Reply by lfarv** [reply]
@carmignani 
Nothing to do, no toolbox: it's pure C, so it's just a compilation flag. The behaviour described for python [here](https://at.readthedocs.io/en/latest/p/howto/multiprocessing.html#installation) is also valid for Matlab: OpneMP is active as soon as the number of particles is equal or more that 10.  The preliminary installation step for MacOS also applies.
But I agree, this should be duplicated in the Matlab section…

**Comment by lfarv** [comment]
#905 allows giving a seed for random generators when starting a tracking in `Lattice.track` (python), `ringpass` and `linepass` (Matlab)
